sql: emit NOTICEs when implicitly dropping indexes

chrisseto opened the pull request:
**sql: emit NOTICEs when implicitly dropping indexes**
Previously, users would have to either know ahead of time or later
discover if an index would be implicitly dropped by dropping a column
that the index referenced.

At best, this was mildly surprising. At worst, workloads could be
dramatically impacted without much indication as to why.

To ensure implicit drops do not go unnoticed, this commit updates `ALTER
TABLE ... DROP COLUMN ...` to emit NOTICEs for each implicitly dropped
index.

Fixes: #106907
Release note (sql change): NOTICEs are now emitted for each index
dropped by an `ALTER TABLE ... DROP COLUMN ...` statement.

**sql: warn users that DROP COLUMN may drop indexes**
Previously, `ALTER TABLE ... DROP COLUMN ...` statements would warn
users that all data held by that column would be destroyed when
`sql_safe_updates = on`. Unlike the [documentation for `DROP
COLUMN`](https://www.cockroachlabs.com/docs/stable/alter-table#drop-column),
this message does not indicate that indexes may be dropped.

This commit extends the warning to also inform users that any indexes
referencing that column will also be dropped to match the warnings in
our documentation.

Epic: none
Informs: #106907

Release note: (sql change): Attempting to drop a column when
`sql_safe_updates = on` now additionally warns users that indexes
referencing that column will be automatically dropped.

chrisseto committed:
sql: warn users that DROP COLUMN may drop indexes

Previously, `ALTER TABLE ... DROP COLUMN ...` statements would warn
users that all data held by that column would be destroyed when
`sql_safe_updates = on`. Unlike the [documentation for `DROP
COLUMN`](https://www.cockroachlabs.com/docs/stable/alter-table#drop-column),
this message does not indicate that indexes may be dropped.

This commit extends the warning to also inform users that any indexes
referencing that column will also be dropped to match the warnings in
our documentation.

Epic: none
Informs: #106907

Release note: (sql change): Attempting to drop a column when
`sql_safe_updates = on` now additionally warns users that indexes
referencing that column will be automatically dropped.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108047)
<!-- Reviewable:end -->

chrisseto committed:
sql: emit NOTICEs when implicitly dropping indexes

Previously, users would have to either know ahead of time or later
discover if an index would be implicitly dropped by dropping a column
that the index referenced.

At best, this was mildly surprising. At worst, workloads could be
dramatically impacted without much indication as to why.

To ensure implicit drops do not go unnoticed, this commit updates `ALTER
TABLE ... DROP COLUMN ...` to emit NOTICEs for each implicitly dropped
index.

Fixes: #106907
Release note (sql change): NOTICEs are now emitted for each index
dropped by an `ALTER TABLE ... DROP COLUMN ...` statement.

Xiang-Gu commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -403,12 +405,16 @@ CREATE TABLE o (gf INT REFERENCES t (g), h INT, i INT, INDEX oi (i), INDEX oh (h
+# Example: "NOTICE: the data for dropped indexes is reclaimed
+# asynchronously\nHINT: The reclamation delay can be customized in the zone
+# configuration for the table."
+statement notice (NOTICE: dropping index "(oh|oih)" which depends on column "h"(\nNOTICE: .+\nHINT: .+\n)?\n?){2}
```
The regex here is indeed a bit nasty. Can we just do something like "NOTICE: dropping index "(oh|oih)" which depends on column "h".*"

chrisseto replied:
I agree that it's pretty nasty! I didn't want to `.*` it because I'm a bit worried of "slurping up" too much. Though, `.` doesn't match `\n`'s by default in go. So maybe that's paranoia? WDYT?

rafiss replied:
i'm OK with the test having a tricky regex, especially since you have a good comment for it.

chrisseto commented:
TFTRs!

bors r+
